Optimise status colours and icons. Closes #6174.

magao opened the pull request:
This pull request addresses issue #6174.

Swaps the colours used for the paused downloading and stalled states, and makes the stalled state use maroon for a light theme (which I find easier to read).

The reasoning here is that paused downloading is a "normal" state, whereas stalled is a "warning" state, so the colours should reflect that.

LordNyriox commented:
@Chocobo1:  What do you think?

Chocobo1 commented:
@magao 
Could you post a screenshot?

magao commented:
![qbittorrent_colours](https://cloud.githubusercontent.com/assets/1728015/21598206/69b7e498-d1ac-11e6-8fda-5262591ad027.png)

I'm not sure how to make it use the dark theme colours  - changing my theme in Mint Mate to a dark one didn't appear to do it.

LordNyriox commented:
@sledgehammer999, @Chocobo1:  Do you folks remember who did the revised status colors for v3.0, and the original reasoning behind them?  

Thanks.

LordNyriox commented:
@magao:  Since I have no problems with a solid-black "high contrast" theme on Windows 8.1 with qBittorrent, that certainly might be a good place to start (if you need to test the colors).  Use "High Contrast Black".

magao commented:
Urgh! That requires getting qBittorrent compiling on Windows ... definitely not the trivial task it is on Linux and something I was hoping to avoid.

On the plus side, my Windows machine should be much faster at compiling than my qBittorrent Linux VM.

LordNyriox commented:
@magao:  You can count me out!  If I could figure out how to compile qBittorrent on Windows I would have done it years ago.  But no matter how many times I read the instructions (and try to follow them), I never get past the library compilation.  :\

thalieht commented:
Breeze dark theme
<img width="300" alt="colors" src="https://cloud.githubusercontent.com/assets/6451685/21682110/62ce8136-d35b-11e6-9376-fe3d0c4e9faf.PNG">

magao commented:
Thanks!

Would you be able to change it to use maroon for the dark theme as well so we can compare? I suspect maroon will be too dark.

thalieht commented:
Indeed
<img width="300" alt="capture2" src="https://cloud.githubusercontent.com/assets/6451685/21696954/decc795e-d398-11e6-8f3c-b8b017e8a84e.PNG">

thalieht commented:
You could use something like http://contrastchecker.com/ to test colors if you want.

magao commented:
OK - so I think the colours in the pull request work as they are, so it's just a question of whether the change is desirable or not.

magao commented:
Rebased on top of #6212 and uncrustified.

magao commented:
The new icon scheme #4253 is based on the existing colours - stalled and paused look a little weird in my build. If this PR gets merged then might want to adjust the icon colours to match.

Balls0fSteel commented:
I am pretty sure it's impossible to find a scheme that will fit every distribution's default colors. Maybe allow color customization via the config file, so the distribution packagers can set it up correctly? Then again, that's just another added load of bloat.

magao commented:
Rebased on top of master.

I'll see if I can work out how to change the colours of the icons to match to include in this PR, but I'm not a graphical person ...

Balls0fSteel commented:
@magao The easiest method is using Hue/Saturation in Paint.net/GIMP. (But yeah, see my previous comment, if you match one theme/platform, it will look weird on the other one. Maybe just a new ini for the colors?)

magao commented:
I've gone with a different approach - when loading the icons, set the icon colour to match the state colour. This will handle both light and dark themes. Doesn't require any changes to the icon files - just creates a new solid-colour pixmap masked by the transparent pixels.

![icons](https://cloud.githubusercontent.com/assets/1728015/22309269/ab332f14-e39e-11e6-9f26-541f7f4b50fb.png)

I've also got an alternative that I prefer that I've included in the PR (72399d3) for discussion - inverting the icons. This also doesn't require any changes to the icon files (just inverts the mask).

![icons_inverted](https://cloud.githubusercontent.com/assets/1728015/22309479/8b258860-e39f-11e6-80a7-e6a5378730b1.png)

If you like 1ea2539 and aren't sure about the others yet (or don't want them) I can split 1ea2539 out into its own PR.

LordNyriox commented:
@magao:  Actually that last commit does something suggested by @sledgehammer999 himself:  

<https://github.com/qbittorrent/qBittorrent/pull/4253#issuecomment-183712105> 

Thanks again!

LordNyriox commented:
@magao:  Beautiful work!  

That last commit actually does something suggested by @sledgehammer999 himself:  

<https://github.com/qbittorrent/qBittorrent/pull/4253#issuecomment-183712105>:  

> As for the "skins" folder. I would like for the icons that appear in the sidepanel for the status of torrents(which also appear in the transferlist) to lose their rectangle. Their should have only the symbol which should be colored accordingly(not white).
What do you guys think? 

Thanks again! 
– Nyriox

magao commented:
Interesting - I just realised that my change doesn't invert the status icons in the sidebar (I don't normally display them). I'll work out how to fix that and re-push.

magao commented:
I've got a new series of commits (not pushed to github yet) which I could either leave as a single PR, or split them up. In the options below, if the final one is where we want to end up then I think it would be best to just have everything in this single PR.

For the transfer filter list, I've special-cased `All` to use the same colour as in the icon file (since there's no corresponding torrent state). All other filters use the colour from one of the torrent states. `Resumed` and `Active` use the same colour as `Downloading`, `Inactive` uses the same colour as `Stalled Downloading`.

The following are all on a light theme - the screenshots in earlier commends show what would happen for a dark theme (in particular, `Paused` will be Gray 80 and `Inactive` will be Salmon).

Original state colours, current icons

![issue6174_pre_filtericons](https://cloud.githubusercontent.com/assets/1728015/22318357/eb82eb48-e3cd-11e6-91c0-c2174e594ddf.png)

Modified state colours, current icons

![issue6174_post_filtericons](https://cloud.githubusercontent.com/assets/1728015/22318358/f3e05956-e3cd-11e6-8800-a3c8f11cce70.png)

Original state colours, inverted icons

![issue6174_pre_filterinverticons](https://cloud.githubusercontent.com/assets/1728015/22318366/031e66c4-e3ce-11e6-89f0-cee3eae6cffa.png)

Modified state colours, inverted icons (this is what I'll be using no matter the decision)

![issue6174_post_filterinverticons](https://cloud.githubusercontent.com/assets/1728015/22318368/0c806744-e3ce-11e6-8ac2-4afd31087443.png)

Balls0fSteel commented:
@magao Very nice job. 👍

magao commented:
Note that this doesn't change anything for the web UI - the icon files will need to be changed for that (or an equivalent to what I've done be implemented in javascript).

evsh commented:
While joining the other who like the icons look, I would like to raise a couple of concerns regarding the code in this PR.
1. You, in fact, use a singleton to manage all the icons, but you do it implicitly, and this is not good. One can benefit from a proper implementation. For example, it can be used in the web UI as well. A singleton 
```C++
class Skin {
public:
    QIcon icon(SkinIconEnum icon);
    QIcon stateIcon(TorrentState state);
    QColor stateColor(TorrentState state)
};
```
might be a much cleaner solution. Perhaps you wanted to change the current code as little as possible, but your changes make the code even uglier. Plus, your code can not react to colour theme changes in principle. While now icons are readable when switching from light to dark colour theme (and vise-versa), with these changes reading them would become problematic.   

2. Replacing colours in pixmaps will lead to problems if an icon is rendered with anti-aliasing (you kill it). At least I do see the icons in this PR aliased on 128 DPI screen. It would be much better to replace colours in SVG and render an icon afterwards.

My suggestion, therefore, is to implement the aforementioned `Skin` singleton, edit SVG skin icons to make them monochrome with simple colours (e.g. black on transparent) and then do simple text replacement (without fiddling around with SVG rendering).

LordNyriox commented:
@magao:  As far as fixing the original icons goes, the workflow for icon creation has changed in <https://github.com/qbittorrent/qBittorrent/pull/4253>.  

In particular, the icons have an SVG source format, which is then converted to PNG by a "grunt task".  

I have some experienceworking with vector graphics so I can perform a fix for the original SVG files if you give me a day or so (I need to get Inkscape up and working). 

I don't know how the PNG conversion works, but I'll figure it out shortly.

LordNyriox commented:
@magao:  Here's the entire set of status icons inverted to remove their squares (SVG & PNG):  [<https://mega.nz/#!DNsAXJIa!mjBscen0GTfH7qYgM1Q7g6mZf9OPWJZRtEBsV7CYEi4>] 

For those that care, the process for producing these in Inkscape was 7-step (per icon): 
1.  Use the "dropper" color-select tool on the colored background.  
2.  Use the "bucket" color-fill tool (without selecting anything) on the symbols in the center.  Icons like "paused" and "queued" required separate clicks to the various parts. 
3.  Use the "path-select" tool to click anywhere on the colored background (this selects the background).  
4.  Use the `Edit`->`Delete` tool to delete the (selected) background.  
5.  Repeat steps 3-4 for "inner" background areas.  (For "queued", "checking" etc. icons.)  
6.  Save the resulting SVG file.  
7.  Use `File`->`Export to PNG`.  Set area to `Page` and `Image size` to **256**x**256**.  Then click `Export`.  

Thanks again, folks!  
~ Lord Nyriox

magao commented:
I've worked out how to change the original SVG files to invert them (using masks - thank you text-based graphics format) so I'll do that (and modify the colours to match my new light theme) and then use the instructions to regenerate the PNG files.

@evsh Could you post a screenshot of both the icons on a dark theme, and with anti-aliasing artifacts? I may be able to find a method of doing the fill without the artifacts, which I think is a better option than needing to produce separate dark and light theme icons.

I think making more drastic changes such as how the caching works (i.e. allowing swapping between light and dark themes dynamically) needs to be a separate issue. For example, the transfer filter list is only built once - we would need to detect the theme change and then rebuild the transfer filter list.

If the code is to be refactored, it should probably be done by someone who has a better grasp of the whole picture. For example, I have no idea how the web UI fits in other than the image files being used. That's the main reason I want to have as light a touch as possible.

magao commented:
WebUI with modified icons - inverted and matching new state colours.

![qbt-webui-newicons](https://cloud.githubusercontent.com/assets/1728015/22352950/53e4a4b6-e472-11e6-9ec1-6f5e7b12ecc4.png)

Does the WebUI support dark themes? I did a text search for "dark" and the only thing I could find was in progressbar.js which seems unrelated.

evsh commented:
> @evsh Could you post a screenshot of both the icons on a dark theme, and with anti-aliasing artifacts?

Here is the screenshot, please:
![qbt-new-state-icons](https://cloud.githubusercontent.com/assets/394621/22356236/02cfee7e-e430-11e6-82db-8899c9327b35.png)
I have to admit that I was wrong regarding the antialiasing: the original icons are 256 by 256 pixels and that is enough when downscaling. What I do see is the absence of the sub-pixel rendering (compare with the letters at the right). But that is a completely different story.

P.S. Most of my torrents are in stalled state. Therefore, I would prefer a neutral colour for them, but now it is too reddish.

magao commented:
@evsh I do feel that the Salmon is too light. It used to be used for paused as you can see in my earlier comments, and changing that was the whole point of this PR. I think we need some red in it, and I like the Maroon for the light theme, but we'd need something better for the dark theme then (as can also be seen in earlier comments, Maroon is too dark).

You're probably best placed to come up with a colour you like on a dark theme. The good thing is that with the PR in its current state you can experiment just by modifying `TorrentModel::getColorByState()`.

I'm glad the antialiasing isn't a problem - I was afraid that `QPixmap` was doing something weird when I made the masked solid colour. I'm not surprised there's no sub-pixel antialiasing as that's usually only a feature of the font system (but maybe there's a flag for Qt to do it for icons ... I'll have a look).

magao commented:
@LordNyriox I'll compare your SVG files with mine, and take the simplest (e.g. if yours aren't done with masking, that's probably simpler for someone else to modify in the future).

evsh commented:
My point is: the most common status (stalled) should be mapped into a neutral colour (a shade of grey). I use my own modification (#3936)  for the colour theme cause I don't think a single set of colours can satisfy  everyone. That is why my only suggestion is to be conservative: as little different colours as possible and prefer low saturated ones, especially for elements that fill a significant part of a window.

magao commented:
@evsh Your most common status (stalled) is different to my most common status (paused). In my case, I have a lot of RSS filtering, and I have all torrents paused by default so I can pick and choose (there's often several options available). As you said, no theme is going to satisfy everyone.

Stalled uploading is generally an inconsequential status, and so IMO should be either a neutral colour or the same as active uploading (royal blue on light there, steel blue on dark theme). I'm going to implement having both uploading states be the same in my next version of the PR.

Stalled downloading is almost always undesirable e.g. if all my torrents are stalled downloading for a while it might indicate that my VPN connection has gone down. I think it definitely deserves a warning colour (i.e. reddish).

Note that there are cases that stalled uploading could indicate a problem (e.g. a networking issue).

evsh commented:
@sledgehammer999, would you accept a PR which implements skin colours selection by a user adding a new options page "Appearance" (possibly merging with  #6196)?

evsh commented:
@magao indeed, the most common status may differ from user to user. This complicates selecting a universal theme even more.

LordNyriox commented:
@magao:  Did you read the "how-to" guide I wrote?  I used no masking or fancy features.  I literally just bucket-dumped the background color on the foreground areas, then deleted the background textures.  The method I used was 100% manual object conversion.  

Masks are the one feature of image/vector processing I have never been able to use.

LordNyriox commented:
@magao:  If you can figure out a way to export the paths without manual-editing "gunk", that would be preferred.  I did notice during the conversion process that evidently not **everything** in the background was properly removed.  

Since I don't know enough about Inkscape (it crashed on me 12 times during the icon-fixing process) to fix it myself, I'd thank you for doing so.  

Thanks again, 
~ Nyriox

magao commented:
@LordNyriox Yes - I did but I'd already gone about it another way which I could do with just an SVG viewer and text editor ;) The result is sub-optimal for later graphical editing because it's using masking which is why I want to take yours and minimise the differences.

LordNyriox commented:
@magao:  Might you recommend a (free) lightweight svg editor/viewer?  Inkscape is incredibly complex (and buggy), but it was the only one I could find.  

Thanks again!

LordNyriox commented:
@magao:  I am going to try re-creating the process used for the above SVG files in a different editor today.  Maybe a Chrome app ("Boxy SVG") will do the job better than Inkscape.  

While we are on the subject, what colors (hex please) are you planning on for each icon?  Having a list of colors for each file would allow me to copy the main paths without all the "gunk".  

Thanks again!

magao commented:
@LordNyriox I used [https://github.com/SVG-Edit/svgedit](SVG-edit), but I didn't use any of the graphical features other than viewing the changes. Once I'd worked out the masking I needed for each file I then just modified each SVG in a text editor.

The colours are as per my version of `TorrentModel::getColorByState()` for the light theme (+ the comments above):

checking.svg: #008080
completed.svg: #00008B
downloading.svg: #228B22
error.svg: #FF0000
filteractive.svg: #228B22
filterall.svg: #D1954E
filterinactive.svg: #800000
paused.svg: #000000
queued.svg: #008080
resumed.svg: #228B22
stalledDL.svg: #800000
stalledUL.svg: #4169E1
uploading.svg: #4169E1

magao commented:
@LordNyriox There are definitely some differences between your SVG paths and the originals e.g. checking.svg has less-rounded corners.

Unfortunately, I can't find anything which can take a path and invert it except by masking :(

LordNyriox commented:
@magao:  Actually, the paths were exactly that way in the originals, with corners not-quite-rounded and various other issues.  Those just became more visible after manual inversion.  

Personally, I don't trust masks or any similar automated system for graphics.  Strange as it sounds, I prefer everything done typewriter-style: just enough automation to speed things up, not enough automation to hide data from the user.  

I'm old-fashioned that way.

LordNyriox commented:
@magao:  Something you should note:  

While going through the "skins" directory I found a couple images (which were already there) that actually were already inverted.  I didn't use them in my test build above, but if you copy `download.svg` (which is an inverted form of `downloading.svg`), and `seeding.svg` (which is an inverted form of `uploading.svg`) from the official skins directory and rename the copies, there is 10% of your work right there.  

The only reason I didn't use them before was because of concerns over possible color consistency.  With the recolor I am about to do, that is no longer a concern.  

Thanks again!

LordNyriox commented:
@magao:  Well I just put my editing toolkit through its paces.  ;>  

It took longer than expected, but I fixed the icons!  

To begin with, I replaced my source-forms of `downloading.svg` and `stalledDL.svg` with `download.svg` (<https://github.com/qbittorrent/qBittorrent/blob/master/src/icons/skin/build-icons/icons/download.svg>), and replaced `uploading.svg` and `stalledUP.svg` with `seeding.svg` (<https://github.com/qbittorrent/qBittorrent/blob/master/src/icons/skin/build-icons/icons/seeding.svg>).  I did not notice them earlier because I was so focused on the boxes.    

Then, having identified an invisible (and useless) path object near the lower right-hand corner of both `checking.svg` and `completed.svg` while I was working on the files yesterday, I went ahead and removed it from both files.  (That object was an artifact I identified in the original version of the files, so it didn't come from me.)  

After verifying that the other files were free of such blemishes, I went ahead and used a web-optimizer for SVG files called SVGOMG [<https://jakearchibald.github.io/svgomg/>] to fully optimize the SVG files.  

After the SVG files were artifact-free and gunk-free, I finally used a text editor to replace the colors with the ones described above.  

Voila!  The cleanest SVG files I have ever seen.  

Cleaned/optimized/recolored SVG set link:  [<https://mega.nz/#!DY9xgA5T!VQE_YgbIm67uXIgqbUqsRC8btLpHTgsGqm88aszCnsA>].  

PNG files will be posted shortly.  

~ Nyriox

magao commented:
@LordNyriox No need to post PNGs - I'll just generate them using grunt.

LordNyriox commented:
@magao:  Oops.  Well, they're here now.  :P  

Also, what do you think of the update?  I know they are a little "smoother" than the originals, but they are also more lightweight.

LordNyriox commented:
@magao:  I decided to go ahead and create a Github commit of the changed SVG files, so you can simply merge that into your PR if you want.  See [<https://github.com/LordNyriox/qBittorrent/commit/f956536e79b930819e0e5b1c5eabd5c4e62ae3f3>].  

~ Nyriox

magao commented:
@LordNyriox They look fine to me, except that the downloading and uploading icons are quite a bit bigger now. So obviously there was a reason to have different files. Could you redo `downloading.svg` and `uploading.svg` from the originals?

We'll also need @evsh to check on his hi-DPI monitor once I update the PR (after I get the new `downloading.svg` and `uploading.svg`).

Before:

![issue6174_post_filterinverticons](https://cloud.githubusercontent.com/assets/1728015/22400179/17a126e6-e602-11e6-8c4c-c1561fcebfb0.png)

After:

![issue6174_post_filterinverticons2](https://cloud.githubusercontent.com/assets/1728015/22400181/1bab29ee-e602-11e6-8844-abe8773e1f70.png)

LordNyriox commented:
@magao:  Personally, I find the "bigger" downloading/uploading images more visible (I use a hi-DPI monitor myself) as compared to the version I modified from the "framed" versions.  

As a matter of fact, I actually wanted to remove the excess white-space from the other images, too, but I couldn't figure any way to go about it.  They icons (other than downloading/uploading) are not as visible without the frame due to their size.  

If you compare on a purely objective level those to images you have there, which seems more visible?

LordNyriox commented:
@magao:  If I may ask, can you please update the PR with what I have now first?  My working drive actually is not available right now, so I will not be able to set up the images until tomorrow.  

**NOTE:**
Just so you know, the white-space reduction (for downloading and uploading SVG images) was deliberate.  The reduced framing actually makes the arrows more visible on my 15-inch 1080p screen as compared to my original test images.  

If I could figure out a reasonable way to reduce the framing on the other images, I would.  

Thanks again! 
– Nyriox

magao commented:
@LordNyriox I think it would be best to keep the smaller icons in this PR and then maybe you could create a separate PR for larger icons. This PR has already become much much larger than I originally envisaged (but OTOH I think the result is also much better).

I still need to reorder the PR (need to rebase your commit into it, etc) and I'm working my paid job right now so I might not get to it until the end of the day. So maybe best to wait until I have final versions for all the (small) icons to reduce churn?

LordNyriox commented:
@magao:  All right then.  I actually considered that aspect, so I will update the commit shortly.

LordNyriox commented:
@magao:  As an aside, you probably should change the title of this PR to `Optimize status colors and icons` or something like that.  

Both the icon change and the color shift are visual optimizations, after all.

LordNyriox commented:
@magao:  Updated commit to [<https://github.com/LordNyriox/qBittorrent/commit/629ad803353864427d6adddd946fbe8c15b9c606>].  

Also, there is no "viewport" in the actual text of these files.  There is a "viewbox" however.  Later I'll try to see if there's any way I can manipulate these images by modifying such information.  

Thanks again.

magao commented:
@LordNyriox Thanks - I'll hopefully push a new set of commits later today.

And yes - it is `viewbox` you need to change. The parameters are `x`; `y`; `width`; `height` - you need to reduce `width` and `height` by twice what you increase `x` and `y` to keep the icon centred.

`viewBox="0 0 32 32"` -> `viewBox="2 2 28 28"` -> `viewBox="4 4 24 24"` (screenshot from BoxySVG)

![checking_viewbox](https://cloud.githubusercontent.com/assets/1728015/22441688/8cbeb336-e78c-11e6-8161-fa5472d50e3b.png)

The image remains the same size, but when you then convert to PNG it's the viewbox dimensions that are used, so the icon ends up larger.

Cormak commented:
This and the new theme have me excited for 3.4

magao commented:
Integrated inverted icons from @LordNyriox. I rebased your commit into my series so you may want to delete the branch from your own repository to prevent duplication of changesets (maybe after this PR is merged).

@sledgehammer999 I think this PR is now ready for review. The only thing is the refactoring suggested by @evsh, but as I said I think I'd prefer that to be a separate PR once this is merged (and preferably done by someone who understands the effects on the WebUI as well).

LordNyriox commented:
@magao:  I finally found out how to properly invert the icons!  The explanation here [<http://graphicdesign.stackexchange.com/questions/84271/manually-invert-path-fill-for-svg>] helped me to figure out how the cutouts were created, and exactly reverse the original code--without making unnecessary (manual) edits.  

In short, I used the exact method described in that post for all icons, and for `checking.svg` and `queued.svg` I also removed the separation of the paths (creating the internal gear-hole and clock-face).  

I'll upload a new commit with the cleaner images shortly.

magao commented:
Something happened with https://github.com/LordNyriox/qBittorrent/commit/547bd816514a173a79b3b1f7f87f8e0f983d4986 - they changed from `viewbox="0 0 32 32"` to `width="32" height="32"`. When running `grunt` this resulted in a tiny 32x32 image in the top-right corner of a 256x256 png ...

Fortunately, changing back to using `viewbox="0 0 32 32"` results in restoring the original size, so I'll make that modification in-flight.

magao commented:
Rebased on top of master, integrated new icons from @LordNyriox + my fixes.

@LordNyriox would you like me to squash my fixes from 1b8f521 into your commit?

LordNyriox commented:
@magao:  I noticed a couple of the files had their coordinates rescaled in your fix commit. 

Before you merge the changes (which you are welcome to do), can you rescale all the SVG coordinates to 256x256?  The 32x32 size is actually an artifact from an earlier version of the original icons PR, but I couldn't figure out how to rescale SVG source data.

magao commented:
Does it really matter? `grunt` will produce the same 256x256 PNG regardless of whether the source is 32x32 or 256x256 (since SVG is a vector format). Having them with the same `viewBox` dimensions is more for consistency and ease of experimentation.

I rescaled by resizing the document and object in Inkscape, then manually positioning the object, then modifying the SVG XML to remove all the cruft - a somewhat error-prone process that I'd prefer not to do for all the icons.

magao commented:
Looks like I'm going to include an option for large status icons in this PR after all. They use the original same PNG files, but trim off most of the space around them before converting to an icon (so they rely on the icons having that space). I've also got the status icons dynamically responding to change of dark/light theme.

Need to clean up the changesets before I push - probably this evening.

Mint-X

![small_status_light](https://cloud.githubusercontent.com/assets/1728015/22631713/557b5f4e-ec66-11e6-9488-78ce93a16489.png)

![large_status_light](https://cloud.githubusercontent.com/assets/1728015/22631763/f6c91814-ec66-11e6-9baf-e02235b149c7.png)

Mint-Y-Dark

![small_status_dark](https://cloud.githubusercontent.com/assets/1728015/22631550/6d664c2a-ec63-11e6-90f1-51c3cdc92fa3.png)

![large_status_dark](https://cloud.githubusercontent.com/assets/1728015/22631767/ff424e02-ec66-11e6-81c7-9c82630171a8.png)

magao commented:
@sledgehammer999 @evsh @LordNyriox 

The commits have been refactored and the actual order doesn't match what is showing above in the PR. Look at the [commits page](https://github.com/qbittorrent/qBittorrent/pull/6186/commits) instead (although that still has some weirdness with the _Follow project coding style_ commits - there should only be one at the start).

1. Refactored the icon caching. Most of the status icon handling code has been moved to `GuiIconProvider`. Other components now only need to specify an icon ID e.g. `filterall`, torrent state, or both in a few cases.

2. Torrent and filter lists follow the status colour.

3. Dynamically detects light/dark theme changes and switches to the appropriate icon colours.

4. New preference "Use large status icons". Switches between the small and large icons as above (updates when clicking "Apply").

5. Status colour changes as discussed above.

Bah - just spent an hour trying to work out why I was building successfully, but some of the AppVeyor and Travis builds were failing. Turns out I'd put some code inside an `#ifdef` that it shouldn't have been, and missed putting some inside an `#ifdef` that should have been ...

evsh commented on src/gui/guiiconprovider.h:
```diff
@@ -30,30 +31,44 @@
+        LightTheme = 1,
+        DarkTheme = 2,
+        SmallStatusIcons = 4,
+        LargeStatusIcons = 8,
```
I find this flag and setting name misleading. To me this mode produces border-less or, perhaps, expanded icons.

LordNyriox replied:
@evsh:  `LargeStatusIcons` reduces the whitespace around the icons.  If you look at the screenshots above, it does indeed make the icons look "bigger".

magao replied:
The visible part of the icon is larger. Everything that is removed is transparent.

@sledgehammer999 Your opinion? How would you like the setting named?

evsh replied:
Yes, I understand this. But, you know, a "larger" is pointing to enlarging the object but not changing its internal structure.

LordNyriox replied:
@evsh:  Which is exactly what it does: a realtime autocrop.

LordNyriox replied:
@evsh:  The files are not changed at all.  They are simply rendered without their transparent whitespace (in real-time).

LordNyriox replied:
@evsh, @magao:  Since this option is implemented at the rendering level, would `Render large status icons` work?

evsh commented on src/gui/torrentmodel.cpp:
```diff
@@ -314,47 +298,53 @@ void TorrentModel::handleTorrentsUpdated()
     case BitTorrent::TorrentState::ForcedDownloading:
     case BitTorrent::TorrentState::DownloadingMetadata:
-        return getDownloadingIcon();
+        return "downloading";
```
Could you put `Qlatin1String("literal")` here and below, please?

magao replied:
What's the advantage? It's still going to then be copied to a new `QString` to be returned, so there's no efficiency advantage.

Returning a `QLatin1String` would be incorrect since the `const char *` used to construct it isn't copied - unless we made the literals constants (in which case we could return a `const QString &`). Hmm - could also do that for the colours below.

evsh replied:
`QString` constructor converts literals from UTF-8, while via `Qlatin1String` we specify that the literal is in Latin1 encoding, which is more efficient. When we drop Qt4, we will replace those with `QLatin1Literal`, which will be event more memory efficient.

magao replied:
I've gone with the approach of making the literals `static const QLatin1String` and I can then return a `const QLatin1String &` reference. That could open up other optimisations such as having `GuiIconProvider::getStatusIcon` take a `const QLatin1String &`, but then it would still need to be converted to `QString` internally (can't store `QLatin1String` in a hash as the `const char *` isn't copied). Need to think about this some more.

Chocobo1 replied:
>and I can then return a const QLatin1String & reference...

please don't over-complicate things...
returning QString is perfectly fine, either it will be optimized out by compiler RVO or it will just change QString internal pointer, either way no real copy will be performed (if not modified).

evsh commented on src/gui/torrentmodel.cpp:
```diff
@@ -314,47 +298,53 @@ void TorrentModel::handleTorrentsUpdated()
-QColor getColorByState(BitTorrent::TorrentState state)
+QIcon TorrentModel::getIconByState(BitTorrent::TorrentState state, const QString &overrideIconId)
+{
+    const QString &iconId(overrideIconId.isEmpty() ? getIconId(state) : overrideIconId);
```
This seems to be error-prone to me (the reference). I would merge this line into the next one.

magao replied:
I did that initially but found it to be too complicated. But I thought I'd removed the `&overrideIcon` here as I don't think I use it anymore ... let me check.

evsh replied:
You use a reference to a temporary object, returned from `getIconId()`, don't you?

magao replied:
`overrideIconId` is gone.

evsh commented on src/base/iconprovider.h:
```diff
@@ -43,7 +44,7 @@ class IconProvider : public QObject
     static IconProvider *instance();
 
-    virtual QString getIconPath(const QString &iconId);
+    virtual QString getIconPath(const QString &iconId, const QString &overrideTheme = QString());
```
Don't like this API. And don't like the fact that our icons are split onto two parts while they naturally form the single theme. UNIX users would benefit from merging them together and distributing them as individual files but not resources. Perhaps we better merge these icons into the main theme right away (for instance into a subdir "status", as per FDO specification)? Because the proposed API looks like it allows to mix icons from different themes, which shall not happen and certainly may not be encouraged by an API.

magao replied:
I wasn't intending to change the file layout at all, but there is already a `src/icons/flags/` directory, so there's precedence for doing this.

@sledgehammer999 Do you want me to reorganise so there's a `src/icons/status/` directory?

evsh commented on src/base/preferences.cpp:
```diff
@@ -57,9 +57,11 @@
+Preferences *Preferences::m_instance = 0;
 
-Preferences::Preferences() {}
+Preferences::Preferences()
```
`= default` too?

evsh commented on src/gui/guiiconprovider.cpp:
```diff
@@ -117,12 +129,13 @@ QIcon GuiIconProvider::generateDifferentSizes(const QIcon &icon)
 #endif
 
-QString GuiIconProvider::getIconPath(const QString &iconId)
+QString GuiIconProvider::getIconPath(const QString &iconId, const QString &overrideTheme)
```
Your changes seem to need a rebasing on top of the current master (the signature of this function is old).

magao replied:
This was after rebasing, and I didn't see any changes. Will become moot if @sledgehammer999 wants me to reorganise the icon hierarchy.

evsh replied:
My bad, this is another function.

evsh commented on src/gui/torrentmodel.h:
```diff
@@ -33,15 +34,20 @@
 #include <QAbstractListModel>
+#include <QColor>
+#include <QIcon>
+#include <QPixmap>
```
Alphabetical order?

magao replied:
https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md#8-header-inclusion-order

magao replied:
Oh - `<QList>` is out of order - is that what you meant?

evsh replied:
Yep.

evsh commented on src/gui/torrentmodel.h:
```diff
@@ -33,15 +34,20 @@
+#include <QPixmap>
 #include <QList>
 
+#include "base/bittorrent/torrenthandle.h"
```
You are adding a bunch of libtorrent headers for the sake of what? Those static functions below do not need access to the private members of the class. Then why?

magao replied:
So that `BitTorrent::TorrentState` is a complete type. But I could get around that by the parameter instead being `const BitTorrent::TorrentState &` - then a forward declaration would be enough.

evsh replied:
I was wondering are these function need to be in the class interface. BTW, you may make `TorrentState` an `enum class` with explicit underlying type and then forward declare it, eliminating the include.

magao commented on src/base/iconprovider.cpp:
```diff
@@ -35,7 +36,9 @@ IconProvider::IconProvider(QObject *parent)
 }
 
-IconProvider::~IconProvider() {}
+IconProvider::~IconProvider()
```
I don't understand. This was just a reformatting via uncrustify.

evsh replied:
OK then.

magao replied:
I see what you mean though - c++11 `= default`. Still catching up on c++11 - much of this stuff wasn't around last time I used c++ (~10 years ago). And it's been over 20 years since I hung around on comp.std.c++ :)

magao commented on src/gui/guiiconprovider.cpp:
```diff
@@ -135,13 +148,87 @@ QString GuiIconProvider::getIconPath(const QString &iconId)
+{
+    static int cachedIconFlags = 0;
+    static QHash<QPair<QString, BitTorrent::TorrentState>, QIcon> cache;
+    int theme = getThemeFlags();
```
Because that's how the dark theme was detected previously, and I simply extended it. Didn't know about `QEvent::PaletteChange` - I tried to find some event that would indicate a theme change but didn't know what to look for.

Now I'll see about hooking into it.

magao replied:
Not sure how to hook it. `GuiIconProvider` doesn't have a parent, so it doesn't seem to receive events.

I'm wondering if maybe `mainwindow` should notice `QEvent::ApplicationPaletteChange` and then forward it as a `Preferences::changed()` signal - or maybe a new `Preferences::themeChanged` signal.

evsh replied:
I would catch that event in the application object. Like the `Preferences::themeChanged` signal.

magao replied:
My current prototype does that, except it's caught in `MainWindow::event()` which already existed.

magao commented on src/gui/guiiconprovider.cpp:
```diff
@@ -135,13 +148,87 @@ QString GuiIconProvider::getIconPath(const QString &iconId)
+        return icon;
+
+    QPixmap pixmap(path);
+    QPixmap &iconPixmap(pixmap);
```
Reference to `pixmap` which may be reassigned to `colored` below, rather than unnecessarily copying objects.

magao replied:
However, I introduced a bug when I fixed up the `#ifdef` stuff so it didn't resize ...

evsh replied:
> which may be reassigned to colored below, rather than unnecessarily copying objects. 

This is wrong in C++. One needs a pointer for this kind of things.

magao replied:
Yes - as I updated my comment, I messed this up. A reference is actually usable for it - the problem was referencing (or pointing to) an object that went out of scope ... I've fixed this in my not-yet-pushed version.

evsh replied:
In C++ a reference can not be reassigned, that is what I wanted to say.

magao replied:
Yes - now I think about it further, which I think was happening is that `operator=()` was being called (hence safe, but no performance advantage).

magao commented on src/gui/guiiconprovider.h:
```diff
@@ -30,30 +31,44 @@
+    enum ThemeFlags
+    {
+        LightTheme = 1,
+        DarkTheme = 2,
```
Yep.

magao commented on src/gui/guiiconprovider.h:
```diff
@@ -30,30 +31,44 @@
     static void initInstance();
     static GuiIconProvider *instance();
 
+    int getThemeFlags();
```
Again, didn't know what to look for. Thanks.

magao commented on src/gui/guiiconprovider.h:
```diff
@@ -30,30 +31,44 @@
     QIcon getFlagIcon(const QString &countryIsoCode);
-    QString getIconPath(const QString &iconId);
+    QIcon getStatusIcon(const QString &iconId);
+    QIcon getStatusIcon(const QString &iconId, BitTorrent::TorrentState state);
```
Because the 2 functions need to call the third with different `ignoreState` flags. I could consolidate down to 2 functions, but that's allowing an implementation detail to intrude into the public API.

This is a part of the API I've struggled with. I don't want to add an "not applicable" entry to the `TorrentState` enum, since that's not actually a state. But that means that I don't have a way to say that the icon doesn't actually correspond to a state (required for the `All` filter). Hence the 2 public functions calling the private function.

evsh replied:
Oh, yes! Pardon me, I forgot about the `All` case and overlooked this in the code. Thank you for the explanation.

evsh replied:
BTW, a C++-ish approach to this would be the following:
```C++
struct FuncArgs {
  FuncArgs() {
  // set defaults
  }
  FuncArgs& arg1(Type1 val) {
   m_val1 = val;
   return *this;
  }

  FuncArgs& arg2(Type2 val) {
   m_val2 = val;
   return *this;
  }

private:
    Type1 m_val1;
    Type2 m_val2;
};

void func(const FuncArgs& args = FuncArgs());

func();
func(FuncArgs().arg1(val).arg2(val2));
```

magao replied:
I'm not a fan of that approach - it makes the call sites messy. I think overloads make more sense in this case.

LordNyriox commented:
@sledgehammer999, @glassez, @Chocobo1, @ngosang, @qbittorrent/qbittorrent-frequent-contributors :  
Does anyone have anything to contribute regarding this PR?  It involves some fairly major changes in mainwindow rendering.

LordNyriox commented:
@sledgehammer999, @glassez, @Chocobo1, @ngosang:  
Does anyone have anything to contribute regarding this PR?  It involves some fairly major changes in mainwindow rendering.

magao commented:
@LordNyriox My current prototype removes the direct manipulation of mainwindow and replaces it with capturing the palette change and forwarding a signal. The widgets then connect to that signal and update their icons.

evsh commented:
> My current prototype...

Eager to see it :)

magao commented:
@sledgehammer999 @evsh @LordNyriox @Chocobo1 Rebased on top of master and hopefully addressed all comments (either in discussion or code).

I've included commit 417ec61 which moves the status icons out to their own directory - let me know if you don't want it @sledgehammer999 and I'll remove it from the PR.

I've simplified the API somewhat, and in particular `IconProvider::getIconPath()` no longer takes an optional parameter. Instead, the icon path is hard-coded to `:icons/skin/` (or `:/icons/status/` with 417ec61).

Theme detection now takes place in `GuiIconProvider::configure()`, which emits `GuiIconProvider::themeChanged()` if there actually was a change. This will either be called from `MainWindow::event()` (via `GuiIconProvider::updateTheme()`) in response to a palette change, or due to a `Preferences::changed()` signal (for small/large status icons). Anything which needs to be notified of a theme change simply connects to `GuiIconProvider::themeChanged()`.

BTW I've kept the "large" status icon terminology as the current votes are currently 2 in favour (@magao, @LordNyriox) to 1 against (@evsh).

I've also used `static const` `QLatin1String` and `QColor` inside `TorrentModel::getIconId()` and `TorrentModel::getColorByState()`. I think it makes the code more self-documenting (and potentially slightly more efficient, though unlikely to be significant).

At one point this was a simple 6-line PR ...

evsh commented on src/gui/optionsdlg.ui:
```diff
@@ -213,6 +213,16 @@
                 </widget>
                </item>
                <item>
+                <widget class="QCheckBox" name="checkLargeStatusIcons">
```
Does this little tune really belong to the main options page? I would put it into the advanced page (where we have some icons-related settings already).

magao replied:
I thought about that, but the transfer list UI options seemed like an obvious place for it. It's about as significant as hiding zero and infinity values IMO.

evsh commented on src/gui/torrentmodel.h:
```diff
@@ -33,15 +34,19 @@
 #define TORRENTMODEL_H
 
 #include <QAbstractListModel>
+#include <QColor>
```
All these new includes are unneeded. You can forward declare these classes and move the includes into the .cpp file.

magao replied:
I was thinking that because they were being returned from functions they needed to be complete types. But you can declare functions which accept or return incomplete types. Rusty C++.

https://stackoverflow.com/questions/553682/when-can-i-use-a-forward-declaration

evsh commented on src/gui/transferlistfilterswidget.h:
```diff
@@ -88,6 +90,8 @@ private slots:
     virtual void handleNewTorrent(BitTorrent::TorrentHandle *const);
     virtual void torrentAboutToBeDeleted(BitTorrent::TorrentHandle *const);
+
+    int m_currentThemeFlags;
```
Unused

magao replied:
Whoops - forgot to remove this.

evsh commented:
> At one point this was a simple 6-line PR ...

BTW, there is an unanswered question from that times: this PR floods the main window with a lot of bright colours, effectively colourising all the strings in the torrent list, not only making them hard to read, but also making the main window look tasteless. 

I again suggest to colourise only the torrents that are in an active state which implies quick changes, i.e. those under checksum computations and maybe actively downloaded and the ones with errors. For the rest of them leave a neutral text and colourise icons only.

magao commented:
> BTW, there is an unanswered question from that times: this PR floods the main window with a lot of bright colours, effectively colourising all the strings in the torrent list, not only making them hard to read, but also making the main window look tasteless.

In my main window, it massively reduces the amount of colour as all the paused torrents go from salmon -> black.

I changed StalledUploading to be the same colour as Uploading - did that not fix your problem? If not, what state are the torrents in that you're complaining about?

magao commented on src/gui/guiiconprovider.cpp:
```diff
@@ -27,23 +28,38 @@
+#include <QApplication>
+#include <QBitmap>
+#include <QDebug>
+#include <QEvent>
```
Leftover from my prototype.

magao commented on src/gui/guiiconprovider.h:
```diff
@@ -30,35 +31,68 @@
+
 #include "base/iconprovider.h"
 
+class QEvent;
```
Leftover from my prototype.

sledgehammer999 commented:
Let me say, that I didn't read any of this while it happened. So the discussion is pretty huge for me. I skimmed through it and I realize that it involves changes to font colors, icons(shape and color) and skins?

So @magao can you give me a summary on what is happening here?

magao commented:
As the PR currently stands:

1. Torrent state colour changes (the original PR with some modifications):

    Paused -> black/gray 80
    Stalled Downloading -> maroon/salmon
    Stalled Uploading -> royal blue/steel blue 1 (same as uploading)

    The idea is that stalled downloading indicates an undesirable state (e.g. you're not receiving data, so may wish to switch to a different torrent), whereas paused is a normal state and stalled uploading probably just indicates no one leeching (not sure if that's the case though).

2. Torrent status icons have been inverted - [screenshots](https://github.com/qbittorrent/qBittorrent/pull/6186#issuecomment-275296895)  (thought the actual method of doing it has changed since).

3. Recoloured the icon files to match the new light theme states so they display correctly in the web UI - [screenshot](https://github.com/qbittorrent/qBittorrent/pull/6186#issuecomment-275533917).

4. Torrent status icons are modified in-memory to be the same colour as the text - [screenshot](https://github.com/qbittorrent/qBittorrent/pull/6186#issuecomment-275237621) (though doesn't actually show the text).

5. Dark/light theme is detected and the icons switched to the corresponding colours automatically (the text was already changing) - [screenshot](https://github.com/qbittorrent/qBittorrent/pull/6186#issuecomment-277566991).

6. Option in the preferences to display larger version of the status icons (mainly for people with hi-DPI screens). The icons are modified in-memory by trimming off excess transparent space. Same screenshot as (5).

Note: I've though about reversing this - make the icons files be the larger size, then add transparent space in memory. That way we wouldn't be relying on the amount of excess space in the icon files. Thoughts?

magao commented:
@sledgehammer999 @evsh @LordNyriox @Chocobo1 Rebased on top of master. Addressed review comments. Top two commits are for discussion:

beb94e9: status icons in own directory
4a571de: all text in torrent list black (light theme) or gray 80 (dark theme).

sledgehammer999 commented:
Quick review from your https://github.com/qbittorrent/qBittorrent/pull/6186#issuecomment-278469992:
I think I like the icon changes. The font color changes, I am not so sure.
In any case, I want to focus on some other PRs this week and the next one. So don't wonder why I won't comment on this PR.

magao commented:
If I create the _Appearance_ options tab as discussed in PR [#6196](https://github.com/qbittorrent/qBittorrent/pull/6196#issuecomment-278484599) then the issue could become moot by having options to override the colours.

sledgehammer999 commented:
>If I create the Appearance options tab as discussed in PR #6196 then the issue could become moot by having options to override the colours.

True. So we can keep the current defaults and user can change them.

magao commented:
> True. So we can keep the current defaults and user can change them.

In that case do you want me to resubmit this PR without the colour change, and if so do you want the status icons in their own directory? I think at that point it should be ready for merging.

sledgehammer999 commented:
>In that case do you want me to resubmit this PR without the colour change, and if so do you want the status icons in their own directory? I think at that point it should be ready for merging.

For the moment keep it as it is. Let's see how the Appearance tab implementation will go.

evsh commented:
> 4a571de: all text in torrent list black (light theme) or gray 80 (dark theme).

To me, the new look is superb (with smaller icons)!

Cormak commented:
Will this be ready for 3.4?

LordNyriox commented:
@magao:  FYI, a little over a year ago, @sledgehammer999 changed the text colors used to indicate status.  

I only just now managed to hunt down the original commit:  [<https://github.com/qbittorrent/qBittorrent/commit/276ca0ab8463433daeec473746fe38ad03c14e14>].

LordNyriox commented:
@Cormak:  It is exceedingly unlikely at the moment—it will probably need to be reworked after [<https://github.com/qbittorrent/qBittorrent/pull/6698>] is merged.  

And, while there is significant hope for [#6698] to be merged in time for v3.4, there is also the possibility that it will not.

glassez commented:
What is the status of this PR? Is it still relevant?
